### PR TITLE
[Enhancement] Use ThreadLocalRandom and try-with-resource

### PR DESCRIPTION
**Pull request by satansk**
## What is the purpose of the change

Small improvments, including:

* use `ThreadLocalRandom` instead of `Random`
* use try-with-resource to avoid explicit resource management
* fix duplicate code

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 1]
```diff
@@ -231,21 +216,31 @@ public static InetAddress getLocalAddress() {
...
     }
 
+    private static Optional<InetAddress> toValidAddress(InetAddress address) {
+        if (address instanceof Inet6Address) {
+            Inet6Address v6Address = (Inet6Address) address;
+            if (isValidV6Address(v6Address)) {
+                return Optional.ofNullable(normalizeV6Address(v6Address));
+            }
+        }
+        if (isValidAddress(address)) {
```
Comment (khanimteyaz) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: Better to refactor the method name isValidAddress to **isValidV4Address**. What do you say?

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 2]
```diff
@@ -231,21 +216,31 @@ public static InetAddress getLocalAddress() {
         return localAddress;
     }
 
+    private static Optional<InetAddress> toValidAddress(InetAddress address) {
```
Comment (satansk) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: `Optional` requires some costs regarding preformance, I'm not sure we should encourage it or not in Dubbo.

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 3]
```diff
@@ -30,47 +30,41 @@
...
-
     private static final Logger logger = LoggerFactory.getLogger(NetUtils.class);
-    private static final int RND_PORT_START = 30000;
 
+    // returned port range is [30000, 39999]
+    private static final int RND_PORT_START = 30000;
     private static final int RND_PORT_RANGE = 10000;
 
-    private static final Random RANDOM = new Random(System.currentTimeMillis());
+    // valid port range is (0, 65535]
```
Comment (CrazyHZM) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: The annotation of the attribute is best used
`/**
 * 
 */`

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java (hunk 2 above)
Comment (carryxyh) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: This is yet to be decided. Personally think that you can use the judgment of null. As we have seen, most frameworks internally support 1.8, but rarely use optinal.

Comment (satansk) [reply dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java]: As we discussed in mail list, the community encourage us to use `Optional` :)

**Comment by beiwei30** [comment]
LGTM too.

**Pull request merged**